core/corehttp/readers.go

```diff
@@ -0,0 +1,134 @@
+package corehttp
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/ipfs/go-ipfs/commands/files"
+	"github.com/ipfs/go-ipfs/importer"
+	chunk "github.com/ipfs/go-ipfs/importer/chunk"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	ft "github.com/ipfs/go-ipfs/unixfs"
+)
+
+// helper functions for the gateway handler
+
+func (i *gatewayHandler) newDagFromReader(r io.Reader) (*dag.Node, error) {
+	// TODO(cryptix): change and remove this helper once PR1136 is merged
+	// return ufs.AddFromReader(i.node, r.Body)
+	return importer.BuildDagFromReader(
+		r, i.node.DAG, chunk.DefaultSplitter, importer.BasicPinnerCB(i.node.Pinning.GetManual()))
+}
+
+// add file, or directory recursively
+func (i *gatewayHandler) addFileRecursive(file files.File) (*dag.Node, error) {
+	if file.IsDirectory() {
+		return i.addDirRecursive(file)
+	}
+	return i.newDagFromReader(file)
+}
+
+// add a directory recursively
+func (i *gatewayHandler) addDirRecursive(file files.File) (*dag.Node, error) {
```

Comment:
> (I've taken the opinion that if you send a tar ball, the whole thing will get added, hidden files included, and will not get pinned).
> So, do you think we should add params for dealing with hidden files and pinning to the gateway?

yeah, we've been thinking of adding a `--pin` option to both `ipfs add` and `ipfs object put` (defaulted true for add, false for object put). 

maybe gateway can follow same convention as API (`?<opt-key>=<opt-val>`)

> That would be surprising to me if add in commands does pinning, but doesn't when through gateway.
Similarly for the case of hidden files.

Agreed. `add` should default to the same in both cases.

> It would seem to be to be bad manners to have that kind remote gateway pin everything that anyone adds, and hence keep it all locally.

Yep, agreed. maybe a config can disallow default gateway auto-pin. and result can be returned in a header or in the output.

> My understanding of the gateway is that it is meant to be a service offering a conduit to the ipfs network, not a fully featured endpoint.

Indeed. though users also use their own http gateways for local apps. eventually we'll want some capability based auth --- i.e. add works if you have the right token.